File: binary_search/0704_binary_search/0704_binary_search.go
package main 

import (
	"fmt"
)

func search(nums []int, target int) int {
	l, r  := 0, len(nums) - 1

	for l <= r{
		m := (l + r) / 2
		switch {
		case nums[m] < target:
			l = m + 1
		case target < nums[m]:
			r = m - 1
		default:
			return m
		}
	}

	return -1
}

func main() {
	nums := []int{-1,0,3,5,9,12}
	target := 9
	res := search(nums, target)
	fmt.Println(res)
}

File: binary_search/0704_binary_search/0704_binary_search.py
class Solution(object):
    def search(self, nums, target):
        l, r = 0, len(nums)-1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] < target:
                l = mid + 1
            elif nums[mid] > target:
                r = mid - 1
            else:
                return mid
        return -1


nums = [-1,0,3,5,9,12]
target = 9
res = Solution().search(nums, target)
print(res)